File: internal/collector/collector.go
package collector

import (
	"fmt"
	"strings"
	"time"

	"github.com/curtis1215/cockpit/internal/inventory"
	"github.com/curtis1215/cockpit/internal/sources"
	"github.com/curtis1215/cockpit/internal/store"
	"github.com/curtis1215/cockpit/internal/version"
)

type FetchFunc func(inventory.Software) (sources.SourceResult, error)
type TranslateFunc func(raw string) (string, error)
type Report struct {
	Software       string `json:"software"`
	CurrentVersion string `json:"current_version"`
}

func DefaultFetch(sw inventory.Software) (sources.SourceResult, error) {
	return sources.FetchLatest(sw, nil)
}

func RefreshUpstream(s *store.Store, inv inventory.Inventory, fetch FetchFunc, translate TranslateFunc) {
	for _, sw := range inv.Software {
		latest, err := fetch(sw)
		if err != nil {
			s.AddEvent("error", sw.Name, "", "fetch failed: "+err.Error())
			continue
		}
		existing, _ := s.GetVersion(sw.Name, latest.Version)
		zh := existing.ChangelogZh
		if zh != "" {
			s.AddVersion(sw.Name, latest.Version, "", latest.ChangelogRaw, zh)
			s.SetTranslateStatus(sw.Name, latest.Version, "ready", "")
			continue
		}
		s.AddVersion(sw.Name, latest.Version, "", latest.ChangelogRaw, "")
		if strings.TrimSpace(latest.ChangelogRaw) == "" {
			s.SetTranslateStatus(sw.Name, latest.Version, "none", "")
			continue
		}
		s.SetTranslateStatus(sw.Name, latest.Version, "translating", "")
		out, terr := translate(latest.ChangelogRaw)
		if terr != nil || strings.TrimSpace(out) == "" {
			msg := "empty translation"
			if terr != nil {
				msg = terr.Error()
			}
			// 只改 status，不動 zh——避免與並行 retry 競態時清掉剛寫好的中文。
			s.SetTranslateStatus(sw.Name, latest.Version, "failed", msg)
			s.AddEvent("error", sw.Name, "", fmt.Sprintf("translate failed (raw %d bytes): %s", len(latest.ChangelogRaw), msg))
			continue
		}
		s.UpdateTranslateResult(sw.Name, latest.Version, out, "ready", "")
	}
}

func ApplyVersionReport(s *store.Store, machine string, reports []Report) int {
	now := time.Now().UTC().Format(time.RFC3339)
	applied := 0
	for _, r := range reports {
		if r.Software == "" {
			continue
		}
		latest, _ := s.LatestVersion(r.Software)
		status, _ := version.Compare(r.CurrentVersion, latest.VersionStr)
		s.UpsertInstall(r.Software, machine, r.CurrentVersion, status, now)
		s.AddEvent("check", r.Software, machine, fmt.Sprintf("current=%s latest=%s status=%s", r.CurrentVersion, latest.VersionStr, status))
		applied++
	}
	return applied
}

File: cockpit/collector.py
from __future__ import annotations
from datetime import datetime, timezone

from cockpit import db
from cockpit.models import Inventory
from cockpit.sources import fetch_latest as _default_fetch
from cockpit.translate import translate_changelog as _default_translate
from cockpit.version_parse import compare


def refresh_upstream(conn, inv: Inventory, *, fetch=_default_fetch,
                     translate=_default_translate) -> None:
    """server 端：抓每個 software 的上游最新版 + 翻譯 changelog，存入 versions。"""
    for sw in inv.software:
        try:
            latest = fetch(sw)
        except Exception as e:
            db.add_event(conn, "error", sw.name, None, f"fetch failed: {e}")
            continue
        existing = db.get_version(conn, sw.name, latest.version)
        zh = existing["changelog_zh"] if existing else None
        if zh is None and latest.changelog_raw:
            zh = translate(latest.changelog_raw)
        db.add_version(conn, sw.name, latest.version, None, latest.changelog_raw, zh)


def apply_version_report(conn, inv: Inventory, machine: str, reports) -> int:
    """agent 回報目前版：對每筆比對已知上游最新版、寫 installs。回寫入筆數。"""
    now = datetime.now(timezone.utc).isoformat()
    applied = 0
    for r in reports:
        software = r.get("software")
        current = r.get("current_version")
        if not software:
            continue
        latest = db.get_latest_version(conn, software)
        latest_version = latest["version"] if latest else None
        status, _ = compare(current, latest_version)
        db.upsert_install(conn, software, machine, current, status, now)
        db.add_event(conn, "check", software, machine,
                     f"current={current} latest={latest_version} status={status}")
        applied += 1
    return applied
